### ISSUE TITLE
atomWithObservable unmounts unexepectedly

### ISSUE THREAD

**Issue by dakom** [pmndrs/jotai#2030]
## Summary

`atomWithObservable` unmounts unexepectedly when there is some async dependency. This is particularly bad when the observable is set via some loop, and this bug therefore _also_ affects `jotai-tanstack-query` when `refetchInterval` is used.

It does not happen if there is no async dependency (See the "FIXME" note in the reproduction on line 18)

## Link to reproduction

https://codesandbox.io/s/angry-bassi-r4326h?file=/src/App.tsx 

See the console log, it gets spammed with mount/unmount/mount every tick

## Additional notes

1. @dai-shi pointed out that the `timeMount.onMount` callback on line 27 _never_ gets called
2. this does _not_ occur in a simple reproduction with regular atoms (https://codesandbox.io/s/bitter-bird-yh9dm5?file=/src/App.tsx), so it may be a bug in just `atomWithObservable`

## Check List

Please do not ask questions in issues.

- [x] I've already opened a [discussion](https://github.com/pmndrs/jotai/discussions) before opening this issue, or already discussed in other media.

Please include a minimal reproduction.

- [x] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

**Comment by dai-shi** [comment]
This doesn't cause the behavior either:
```js
  const [_, time] = await Promise.all([get(timeDepAtom), get(timeTriggerAtom)]);
```

---

I would not consider this a high priority, but am very curious how this happens. If anyone is interested, please feel free to dig into it.

**Comment by dakom** [comment]
yeah, interesting...

just to clarify - is this also not a high priority in usecases which depend on frequent ticks? i.e. is mount/unmount/mount extremely cheap and almost like a noop?

**Comment by dai-shi** [comment]
> 1. @dai-shi pointed out that the `timeMount.onMount` callback on line 27 _never_ gets called

Never mind. I just forgot that `onMount` is only available with writable atoms. (TS should complain.)

**Comment by dakom** [comment]
hmmm, using `Subject` for the observable instead of `BehaviorSubject` also works: https://codesandbox.io/s/goofy-roman-x9476g?file=/src/App.tsx

**Comment by dai-shi** [comment]
> just to clarify - is this also not a high priority in usecases which depend on frequent ticks? i.e. is mount/unmount/mount extremely cheap and almost like a noop?

In my mental model, mount/unmount isn't very cheap compared to read/write. In practice, it's probably cheap enough.

**Comment by dai-shi** [comment]
> hmmm, using `Subject` for the observable instead of `BehaviorSubject` also works: https://codesandbox.io/s/goofy-roman-x9476g?file=/src/App.tsx

🤯

**Comment by dakom** [comment]
okay, I've narrowed it down a bit more, it's whenever there's an initial value set. In BehaviorSubject it's nuanced because it comes through the promise firing right away (and will actually fire twice), but this can also happen if you use a regular Subject and just set an initial value: https://codesandbox.io/s/frosty-shannon-2t9677?file=/src/App.tsx

**Comment by dakom** [comment]
I tried attacking this from various angles... I was able to get this test to pass with a rewrite of `atomWithObservable`, but it broke everything else (see [here](https://github.com/dakom/jotai/blob/5da16f0b47d5644ac8ab054ac7a581ff5c152d1c/src/vanilla/utils/atomWithObservable.ts) if you're curious)

I tried a couple other techniques too. Long story short, I'm currently thinking that at least one of these is true:

1. atoms should not recurse on themselves, full stop. This applies to the "mutual recursion" trick too[*]. If any sort of looping is needed, it _must_ be driven from the outside (e.g. via a store subscription).

2. there is something broken in the mount/unmounting logic somewhere and it's causing atoms to unmount too early

I'm curious to hear your thoughts, but fwiw without digging into the details of how the store/dependencies/mounting works - after playing with this code some more, I think it fundamentally feels wrong to recurse in an atom and so I'm leaning towards 1 being the culprit. Although atomWithObservable is super nice, if it's fundamentally unsound, `useObservableAsAtom()` or something that registers the observable subscription in a hook is not that bad.

as for jotai-tanstack-query (and any other libs that depend on this, if any), yeah, this is bad news... but there's a bit of funkiness with that approach as it is (i.e. the whole mechanism of caching between components isn't beneficial at all, jotai already solves that by sharing state between atoms). I think something like `jotai-axios-query` which does a single fetch, combined with a `useRefetchInterval()` effect to drive it is more sound.

Wondering what you think! I may be completely off here...

[*] the mutual recursion trick I'm referring to is where a `write` variable is lifted into global scope, and then a read/write atom sets this variable on mount, but its read function depends on another atom that calls the write. That's a mouthful, but it's what's happening in `atomWithObservable`. In my [broken attempt](https://github.com/dakom/jotai/blob/5da16f0b47d5644ac8ab054ac7a581ff5c152d1c/src/vanilla/utils/atomWithObservable.ts) I made this a bit more obvious by having a third atom tie the loop together, but it's the same idea

**Comment by dai-shi** [comment]
There's still something I'm not 100% following, but here's some thoughts.

Basically, `atomWithObservable` is a hack. More precisely, it doesn't follow the Jotai principle, which is `read` function must be "pure" (= everything must be garbage collectable).
`atomWithObservable` violates that principle, because we can't peek the next value (it requires subscription). That said, it's fairly stable for typical use cases as we have been improving the implementation.

> atoms should not drive themselves, full stop. If any sort of looping is needed, it must be driven from the outside (e.g. via a store subscription).

Can you clarify what's atoms deriving themselves?

> there is something broken in the mount/unmounting logic somewhere and it's causing atoms to unmount too early

There's such a chance that mount/unmounting logic has bugs. It's pretty tricky especially with dependency atoms, and I'm not 100% confident with current code.

**Comment by dakom** [comment]
Thank you for sharing your insight, it's very helpful!

> Can you clarify what's atoms deriving themselves?

Pretty much exactly what you said in other words :) But, I'd say that even [this example](https://github.com/pmndrs/jotai/discussions/2018#discussioncomment-6357244) is problematic:

```typescript
const actionAtom = atom(null, (get, set, arg: string) => {
  console.log(`doing something with ${arg}`);
});
const timerAtom = atom(null, (get, set) => {
  setInterval(
    () => set(actionAtom, "next arg"),
    500
  );
});
```

I don't mean to take out out of the context of that discussion where it was only meant to illustrate the point of separating atoms, I just mean for real-world use where you want to clean up the timer, you only want one instance of it running at a time, etc., I think it's a footgun (if you disagree let me know please!)

I think generally the right way to use atoms as that the `write` semantics should always **originate** from the outside world into jotai. Once you're in the world of jotai atoms, it's fine to use `write` from atom to atom as a compositional tool, but most of the time you should really be `read` and just deriving one piece of data from another

I don't know the deep details, but I do think these issues are highly related to the difficulty of writing FRP libraries. I know it was dealt with extensively in the design of [futures-signals](https://docs.rs/futures-signals/latest/futures_signals/tutorial/index.html#signal) with the separation of `Mutable` (which is like a write atom) and `Signal` (which is like a read atom). Different library and different API (i.e. I sortof lied, you can read from a Mutable, but, in the context of our discussion here it's more read-only) - but I think the idea of "mutating" vs. "transforming" being totally separate needs is relevant and very very hard to get 100% ergonomic and correct

I'm loving jotai btw, very happy to settle into a pattern of best practices and avoid the usage patterns that are less recommended, helps a lot to have clarity on what they are :)

**Comment by dai-shi** [comment]
Yeah, I knew that `setInterval` is problematic. It can return a clean up function and delegate the responsibility to the caller.

In general, you can put impure code in `write` such as mutating variables outside Jotai and communicating with servers, under your responsibility.

**Comment by dai-shi** [comment]
@tmkx Would you be interested in looking into this and/or #2007 (and #2008) ?

**Comment by tmkx** [comment]
```ts
const timeTriggerObservable = new BehaviorSubject(0);
const timeTriggerAtom = atomWithObservable(() => timeTriggerObservable);

timeTriggerAtom.onMount = () => {
  console.log("mount", performance.now());
  return () => console.log("unmount", performance.now());
};

export const timeDepAtom = atom(async () => "timeDep");

const timeAtom = atom(async (get) => {
  //FIXME: not awaiting this fixes the issue
  //i.e. this works (or just not using at all)
  //const _unusedDep = get(timeDepAtom);
  const _unusedDep = await get(timeDepAtom); // 😈 oops
  const time = await get(timeTriggerAtom);
  return time;
});
```
The `timeTriggerAtom` is stuck in a mount/unmount loop because when we await `get(timeDepAtom)`, the `timeAtom` does not depend on `timeTriggerAtom`. As a result, `timeTriggerAtom` keeps getting unmounted.

So, should we unmount dependencies **after** the promise is fulfilled? (currently it's working in a sync manner, so it will clean dependencies when it encounters with the first `await` statement)

**Comment by dai-shi** [comment]
Thanks for the investigation!

> should we unmount dependencies after the promise is fulfilled?

Oh, that sounds too tough. A promise may never fulfill.
Do you think we can fix it with atomWithObservable impl?
Or, is this easily reproducible with bare async atoms?

Using Promise.all can be a workaround, right? If we can't solve it, we could document the workaround.

**Comment by dai-shi** [comment]
ref: https://fosstodon.org/@daishi/111020604426323401

Okay, I think it's a doc issue then. Would you or anyone like to work on it?
